sqlmesh/core/snapshot/cache.py

```diff
@@ -46,15 +49,37 @@ def get_or_load(
         if snapshot_ids_to_load:
             loaded_snapshots = loader(snapshot_ids_to_load)
             for snapshot in loaded_snapshots:
-                self._update_node_hash_cache(snapshot)
-                self.put(snapshot)
                 snapshots[snapshot.snapshot_id] = snapshot
 
+        if c.CAN_FORK:
+            with optimized_query_cache_pool(self._optimized_query_cache) as executor:
+                for key, entry_name in executor.map(
+                    load_optimized_query_cache,
+                    (
+                        (snapshot.model, s_id)
+                        for s_id, snapshot in snapshots.items()
+                        if snapshot.is_model
+                    ),
+                ):
+                    if entry_name:
+                        self._optimized_query_cache.with_optimized_query(
+                            snapshots[key].model, entry_name
+                        )
+
+        for snapshot in snapshots.values():
+            self._update_node_hash_cache(snapshot)
+
+            if snapshot.is_model and not c.CAN_FORK:
+                self._optimized_query_cache.with_optimized_query(snapshot.model)
+
+            entry_name = self._entry_name(snapshot.snapshot_id)
```

Comment:
Why not keep `self.put` on line 50 and remove these 3 lines?

Reply:
because you can't compute full_depends_on until after you've loaded the optimized query